pkg/apis/core/types.go

```diff
@@ -3955,6 +3955,15 @@ type LoadBalancerIngress struct {
 	// +optional
 	Hostname string
 
+	// IPMode specifies how the load-balancer's IP behaves.
+	// Setting this to "VIP" indicates that the traffic passing through
+	// this load-balancer is delivered with the destination IP and port set to the load-balancer's IP and port.
+	// Setting this to "Proxy" indicates that the load-balancer acts like a proxy,
+	// delivering traffic with the destination IP and port set to the node's IP and nodePort or to the pod's IP and targetPort.
+	// This field can only be set when the ip field is also set, and defaults to "VIP" if not specified.
```

Comment:
`Proxy` basically describes the behavior for when `Hostname` is set by `IP` is not, right? I think it would be better to say that:
- it defaults to `Proxy` if `IP` is not set
- it defaults to `VIP` if `IP` is set
- it can't be `VIP` if `IP` is not set

(This would also allow for clear semantics when both `Hostname` and `IP` are set in the future.)

Reply:
Can there ever be a case where the value here is misleading, because a cloud provider is compatible with v1 (generally available) Service and that code is unaware of the defaulting mechanism?

If so, let's work out how we'll document that. End users need to know about places where the API can say one thing and the truth is another, especially if that behavior arrives in a minor release update to a stable API.

We should have an answer to the docs question before we graduate this feature to beta.

Reply:
No; the defaulting behavior described here matches the current semantics: if the cloud LB sets `IP` then kube-proxy will assume it wants VIP behavior, and if it doesn't set `IP` then kube-proxy will assume it wants proxy behavior.

Reply:
I don't understand this description well.

Proxy means "internal cluster traffic directed to the loadbalancer is sent to the external loadbalancer because the loadbalancer is expected to act as a proxy" and VIP is "traffic directed to the loadbalancer inside the cluster is sent to the VIP directly instead of sending it back to the external loadbalancer, no?

Reply:
@aojea You are right.  "Proxy" means the traffic will actually pass through the external loadbalancer and "VIP" means  that kube-proxy will create a forwarding rule on the node using the ingress.ip, directly forwarding the traffic to the backend pods. 
The default value for this field is "VIP" which keeping the current behaviour,  it does not have any impact on the existing cloud providers. BTW, we don't care about the value of ingress.hostname here. @danwinship @sftim 

Comments in the code may not be very clear. I'll try to update it accordingly.

Reply:
We'll (eventually) need docs updates to:
- https://kubernetes.io/docs/reference/networking/virtual-ips/
- https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/
- https://kubernetes.io/docs/tutorials/services/source-ip/

Reply:
I like Dan's thinking, with a tweak:

* it defaults to `"VIP"` when `ip` is set
* otherwise it defaults to `"Proxy"` when `hostname` is set
* otherwise it is not defaulted
* it may not be `"VIP"` if `ip` is not set

Reply:
> I don't understand this description well.

So this can be described either from the CloudProvider's perspective or from kube-proxy's perspective. Probably the docs need to explain both.